File: src/registry.py
"""Schema registry: compile a producer's .proto and derive its table — generically.

Each .proto defines, by convention:
  - GenericFormat   : the common header fields (timestamp, hostname, ...).
  - a payload msg   : the layer-specific fields.
  - a record msg    : { GenericFormat generic_format = 1; <Payload> <name> = 2; }
                      -- one row of data. Identified by having a GenericFormat field.
  - a wrapper msg   : { repeated <record> <name> = 1; }
                      -- what a .pb file actually contains (no delimiters needed;
                      the repeated field structures the records internally).

The loader identifies these by STRUCTURE, not by field names, so a new field, a
renamed payload field, or a new layer needs no code change. The .proto is
compiled at runtime; no generated classes are a build dependency.
"""
from __future__ import annotations

import hashlib
import tempfile
from dataclasses import dataclass
from enum import Enum
from pathlib import Path

import grpc_tools
from grpc_tools import protoc
from google.protobuf import descriptor_pb2, descriptor_pool, message_factory
from google.protobuf.descriptor import FieldDescriptor as FD

from .errors import SchemaError

# Agreed contract with Profile FW: the record message has a singular message
# field named `generic_format` (the header, of type GenericFormat) and a
# singular message field named `payload` (the layer data). Both names are
# required; a .proto that does not follow this is rejected.
GENERIC_FIELD = "generic_format"
PAYLOAD_FIELD = "payload"
GENERIC_MESSAGE = "GenericFormat"

class ColumnType(str, Enum):
    """Database-neutral column types.

    The schema layer speaks only these; each store adapter maps them to its own
    SQL types. Nothing above the store seam names a ClickHouse (or any other
    database's) type.
    """
    STRING = "string"
    BOOL = "bool"
    INT32 = "int32"
    INT64 = "int64"
    UINT32 = "uint32"
    UINT64 = "uint64"
    FLOAT32 = "float32"
    FLOAT64 = "float64"
    DATETIME = "datetime"


# protobuf field type -> neutral column type
_PROTO_TO_TYPE = {
    FD.TYPE_STRING: ColumnType.STRING, FD.TYPE_BYTES: ColumnType.STRING,
    FD.TYPE_BOOL: ColumnType.BOOL,
    FD.TYPE_INT32: ColumnType.INT32, FD.TYPE_SINT32: ColumnType.INT32,
    FD.TYPE_SFIXED32: ColumnType.INT32,
    FD.TYPE_INT64: ColumnType.INT64, FD.TYPE_SINT64: ColumnType.INT64,
    FD.TYPE_SFIXED64: ColumnType.INT64,
    FD.TYPE_UINT32: ColumnType.UINT32, FD.TYPE_FIXED32: ColumnType.UINT32,
    FD.TYPE_UINT64: ColumnType.UINT64, FD.TYPE_FIXED64: ColumnType.UINT64,
    FD.TYPE_FLOAT: ColumnType.FLOAT32, FD.TYPE_DOUBLE: ColumnType.FLOAT64,
    FD.TYPE_ENUM: ColumnType.INT32,
}

# Loader-added bookkeeping columns (not from the proto).
LEADING_COLUMNS = [("run_id", ColumnType.STRING), ("ts", ColumnType.DATETIME)]
TRAILING_COLUMNS = [("_loaded_at", ColumnType.DATETIME)]
# Correlation key, added to the main + child tables ONLY when a payload has
# repeated sub-messages (i.e. child tables exist). A 64-bit unsigned sequence
# assigned per record by the loader; links a main row to its child rows.
# (Deferred to re-architecture: a multi-node-safe scheme — Snowflake/UUIDv7.)
RECORD_ID_COLUMN = ("record_id", ColumnType.UINT64)


@dataclass(frozen=True)
class Column:
    name: str
    type: ColumnType


@dataclass
class ChildSchema:
    """A table for one `repeated <Message>` field in the payload (e.g. per_queue).
    One row per sub-element per record, carrying the header + record_id."""
    table: str                     # <stem>_<repeated_field>
    repeated_field: str            # payload field to iterate (e.g. "per_queue")
    sub_fields: list[str]          # scalar field names of the sub-message
    columns: list[Column]
    order_by: list[str]


@dataclass
class TableSchema:
    stem: str                      # == main table name
    record_cls: type               # the per-row message (StatLog)
    wrapper_cls: type | None       # the repeated wrapper (StatLogs)
    repeated_field: str | None     # the repeated field name inside the wrapper
    generic_field: str             # the GenericFormat field name on the record
    payload_field: str             # the payload field name on the record
    generic_fields: list[str]      # header field names, in column order
    payload_fields: list[str]      # SCALAR payload field names, in column order
    columns: list[Column]          # main table columns, in row order
    order_by: list[str]            # main table ORDER BY
    has_record_id: bool            # True iff children exist
    children: list[ChildSchema]    # one per repeated payload sub-message
    descriptor_bytes: bytes        # picklable, for worker processes
    schema_version: str


def _scalar_columns(msg_desc, source: str) -> tuple[list[str], list[Column]]:
    """Scalar fields of a message -> columns. Rejects repeated/nested — used for
    the header (GenericFormat) and for a repeated sub-message's own fields, both
    of which must be flat scalars."""
    names, cols = [], []
    for f in msg_desc.fields:
        if f.is_repeated:
            raise SchemaError(f"{source}: {msg_desc.name}.{f.name} is repeated "
                              f"(this message's fields must be scalars)")
        if f.type == FD.TYPE_MESSAGE:
            raise SchemaError(f"{source}: {msg_desc.name}.{f.name} is a nested "
                              f"message (this message's fields must be scalars)")
        ct = _PROTO_TO_TYPE.get(f.type)
        if ct is None:
            raise SchemaError(f"{source}: {msg_desc.name}.{f.name} has unsupported "
                              f"protobuf type {f.type}")
        names.append(f.name)
        cols.append(Column(f.name, ct))
    return names, cols


def _split_payload(payload_desc, source: str):
    """Split a payload into scalar fields (-> main table) and repeated
    sub-messages (-> child tables). A repeated scalar or a singular nested
    message is not supported and is rejected with a clear error."""
    scalar_names, scalar_cols, repeated = [], [], []
    for f in payload_desc.fields:
        if f.is_repeated:
            if f.type != FD.TYPE_MESSAGE:
                raise SchemaError(
                    f"{source}: payload.{f.name} is a repeated scalar; only "
                    f"repeated messages (which become child tables) are supported")
            repeated.append(f)          # -> its own child table
            continue
        if f.type == FD.TYPE_MESSAGE:
            raise SchemaError(
                f"{source}: payload.{f.name} is a singular nested message; payload "
                f"fields must be scalars, or repeated messages for child tables")
        ct = _PROTO_TO_TYPE.get(f.type)
        if ct is None:
            raise SchemaError(f"{source}: payload.{f.name} has unsupported "
                              f"protobuf type {f.type}")
        scalar_names.append(f.name)
        scalar_cols.append(Column(f.name, ct))
    return scalar_names, scalar_cols, repeated


def _detect(pool, primary_file, source: str):
    """Find the record message and, if present, the wrapper message.

    The record message is identified by the agreed field NAMES: it has a
    singular message field named `generic_format` (header) and a singular
    message field named `payload` (data). Both are required; anything else is
    a contract violation and is rejected with a clear error.
    """
    messages = {m.name: pool.FindMessageTypeByName(
        f"{primary_file.package + '.' if primary_file.package else ''}{m.name}")
        for m in primary_file.message_type}

    record_desc = generic_field = payload_field = None
    for desc in messages.values():
        by_name = {f.name: f for f in desc.fields}
        g = by_name.get(GENERIC_FIELD)
        if g is None:
            continue   # not a record message

        # this message carries `generic_format`, so it must be a valid record
        if g.type != FD.TYPE_MESSAGE or g.is_repeated:
            raise SchemaError(f"{source}: {desc.name}.{GENERIC_FIELD} must be a "
                              f"singular message field")
        if g.message_type.name != GENERIC_MESSAGE:
            raise SchemaError(f"{source}: {desc.name}.{GENERIC_FIELD} must be of "
                              f"type {GENERIC_MESSAGE}, got {g.message_type.name}")
        p = by_name.get(PAYLOAD_FIELD)
        if p is None:
            raise SchemaError(
                f"{source}: record message {desc.name} has '{GENERIC_FIELD}' but "
                f"no '{PAYLOAD_FIELD}' field. The contract requires both a "
                f"'{GENERIC_FIELD}' and a '{PAYLOAD_FIELD}' message field.")
        if p.type != FD.TYPE_MESSAGE or p.is_repeated:
            raise SchemaError(f"{source}: {desc.name}.{PAYLOAD_FIELD} must be a "
                              f"singular message field")
        if record_desc is not None:
            raise SchemaError(
                f"{source}: more than one record message (both {record_desc.name} "
                f"and {desc.name} have a '{GENERIC_FIELD}' field); expected one")
        record_desc, generic_field, payload_field = desc, g, p

    if record_desc is None:
        raise SchemaError(
            f"{source}: no record message found. The contract requires a message "
            f"with a '{GENERIC_FIELD}' field (header) and a '{PAYLOAD_FIELD}' "
            f"field (data).")

    # wrapper = a message with a `repeated <record>` field
    wrapper_desc = repeated_field = None
    for desc in messages.values():
        for f in desc.fields:
            if (f.is_repeated and f.type == FD.TYPE_MESSAGE
                    and f.message_type.name == record_desc.name):
                wrapper_desc, repeated_field = desc, f.name
                break
        if wrapper_desc is not None:
            break

    return record_desc, generic_field, payload_field, wrapper_desc, repeated_field


def build_schema(stem: str, proto_path: Path) -> TableSchema:
    inc = Path(grpc_tools.__file__).parent / "_proto"
    out = Path(tempfile.mkdtemp()) / "d.desc"
    rc = protoc.main(["protoc", f"-I{proto_path.parent}", f"-I{inc}",
                      f"--descriptor_set_out={out}", "--include_imports",
                      str(proto_path)])
    if rc != 0:
        raise SchemaError(f"{proto_path.name}: protoc failed to compile")
    return build_schema_from_descriptor(stem, out.read_bytes(), source=proto_path.name)


def build_schema_from_descriptor(stem: str, descriptor_bytes: bytes,
                                 source: str = "<descriptor>") -> TableSchema:
    """Derive the table schema from compiled descriptor bytes. Used by both the
    parent (from a .proto file) and worker child processes (bytes are picklable;
    classes are not), so detection rules live in exactly one place."""
    pool = descriptor_pool.DescriptorPool()
    fds = descriptor_pb2.FileDescriptorSet()
    fds.ParseFromString(descriptor_bytes)
    for f in fds.file:
        pool.Add(f)
    primary = next((f for f in fds.file if Path(f.name).name == source), fds.file[-1])

    rec_desc, gfield, pfield, wrap_desc, rep_field = _detect(pool, primary, source)

    record_cls = message_factory.GetMessageClass(rec_desc)
    wrapper_cls = (message_factory.GetMessageClass(wrap_desc)
                   if wrap_desc is not None else None)

    generic_names, generic_cols = _scalar_columns(gfield.message_type, source)
    payload_names, payload_cols, repeated = _split_payload(pfield.message_type, source)

    lead = [Column(n, t) for n, t in LEADING_COLUMNS]
    trail = [Column(n, t) for n, t in TRAILING_COLUMNS]
    rid = Column(*RECORD_ID_COLUMN)
    has_children = len(repeated) > 0

    # main table: run_id, ts, [record_id if children], generics, payload scalars, _loaded_at
    main_cols = lead + ([rid] if has_children else []) + generic_cols + payload_cols + trail

    # child tables: one per repeated sub-message; each carries the header +
    # record_id + the sub-message's scalar fields. Ordered by its key dimension
    # (the sub-message's first field, e.g. queue_id / cpu_id) for fast filtering.
    children: list[ChildSchema] = []
    for f in repeated:
        sub_names, sub_cols = _scalar_columns(f.message_type, source)
        if not sub_names:
            raise SchemaError(f"{source}: payload.{f.name} sub-message has no fields")
        child_cols = lead + [rid] + generic_cols + sub_cols + trail
        _check_unique(child_cols, f"{source}:{f.name}")
        key_dim = sub_names[0]     # convention: first field is the dimension key
        children.append(ChildSchema(
            table=f"{stem}_{f.name}", repeated_field=f.name, sub_fields=sub_names,
            columns=child_cols, order_by=["run_id", "hostname", key_dim, "ts"]))

    _check_unique(main_cols, source)

    return TableSchema(
        stem=stem, record_cls=record_cls, wrapper_cls=wrapper_cls,
        repeated_field=rep_field, generic_field=gfield.name,
        payload_field=pfield.name, generic_fields=generic_names,
        payload_fields=payload_names, columns=main_cols,
        order_by=["run_id", "hostname", "ts"], has_record_id=has_children,
        children=children, descriptor_bytes=descriptor_bytes,
        schema_version="sha256:" + hashlib.sha256(descriptor_bytes).hexdigest()[:16],
    )


def _check_unique(columns: list[Column], source: str) -> None:
    seen = set()
    for c in columns:
        if c.name in seen:
            raise SchemaError(f"{source}: duplicate column '{c.name}' "
                              f"(a generic and a payload/sub field share a name)")
        seen.add(c.name)

File: internal/registry/registry.go
// Package registry compiles a producer's .proto at runtime and derives its
// table schema — generically.
//
// Each .proto defines, by convention:
//   - GenericFormat : the common header fields (timestamp, hostname, ...).
//   - a payload msg : the layer-specific fields.
//   - a record msg  : { GenericFormat generic_format = 1; <Payload> payload = 2; }
//     -- one row of data.
//   - a wrapper msg : { repeated <record> <name> = 1; }
//     -- what a .pb file actually contains (no delimiters needed; the repeated
//     field structures the records internally).
//
// The loader identifies these by STRUCTURE and the agreed field names, not by
// message name, so a new field, a renamed payload message, or a new layer needs
// no code change. The .proto is parsed in pure Go — no protoc binary is
// required at build or run time.
package registry

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/bufbuild/protocompile"
	"google.golang.org/protobuf/reflect/protodesc"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/descriptorpb"

	"github.com/vishal9359/Analysis-FW/internal/fwerr"
	"github.com/vishal9359/Analysis-FW/internal/store"
)

// The agreed contract with Profile FW: the record message has a singular
// message field named generic_format (the header, of type GenericFormat) and a
// singular message field named payload (the layer data). Both names are
// required; a .proto that does not follow this is rejected.
const (
	GenericField   = "generic_format"
	PayloadField   = "payload"
	GenericMessage = "GenericFormat"
)

// protobuf field kind -> neutral column type
var protoToType = map[protoreflect.Kind]store.ColumnType{
	protoreflect.StringKind:   store.TypeString,
	protoreflect.BytesKind:    store.TypeString,
	protoreflect.BoolKind:     store.TypeBool,
	protoreflect.Int32Kind:    store.TypeInt32,
	protoreflect.Sint32Kind:   store.TypeInt32,
	protoreflect.Sfixed32Kind: store.TypeInt32,
	protoreflect.Int64Kind:    store.TypeInt64,
	protoreflect.Sint64Kind:   store.TypeInt64,
	protoreflect.Sfixed64Kind: store.TypeInt64,
	protoreflect.Uint32Kind:   store.TypeUint32,
	protoreflect.Fixed32Kind:  store.TypeUint32,
	protoreflect.Uint64Kind:   store.TypeUint64,
	protoreflect.Fixed64Kind:  store.TypeUint64,
	protoreflect.FloatKind:    store.TypeFloat32,
	protoreflect.DoubleKind:   store.TypeFloat64,
	protoreflect.EnumKind:     store.TypeInt32,
}

// Loader-added bookkeeping columns (not from the proto).
var (
	leadingColumns = []store.Column{
		{Name: "run_id", Type: store.TypeString},
		{Name: "ts", Type: store.TypeDateTime},
	}
	trailingColumns = []store.Column{
		{Name: "_loaded_at", Type: store.TypeDateTime},
	}
	// Correlation key, added to the main + child tables ONLY when a payload has
	// repeated sub-messages (i.e. child tables exist). A 64-bit sequence
	// assigned per record by the loader; links a main row to its child rows.
	// (Deferred to re-architecture: a multi-node-safe scheme.)
	recordIDColumn = store.Column{Name: "record_id", Type: store.TypeUint64}
)

// maxNestDepth bounds descent into singular sub-messages. Protos are finite,
// but a self-referential message would otherwise recurse forever.
const maxNestDepth = 8

// FieldPath locates one scalar value inside a message, descending through
// singular sub-messages. A length-1 path is a direct field; a longer path walks
// a 1:1 nested group (e.g. payload.io_flags.direct).
type FieldPath []protoreflect.FieldDescriptor

// ColumnName is the flattened column name: the field names joined with "_", so
// a nested group keeps its grouping as a prefix (io_flags_direct).
func (p FieldPath) ColumnName() string {
	if len(p) == 1 {
		return string(p[0].Name())
	}
	parts := make([]string, len(p))
	for i, fd := range p {
		parts[i] = string(fd.Name())
	}
	return strings.Join(parts, "_")
}

// ChildSchema is a table for one `repeated <Message>` field in the payload
// (e.g. per_queue). One row per sub-element per record, carrying the header and
// record_id.
type ChildSchema struct {
	Table         string                       // <stem>_<repeated_field>
	RepeatedField protoreflect.FieldDescriptor // payload field to iterate
	SubFields     []FieldPath
	Columns       []store.Column
	OrderBy       []string
}

// TableSchema is everything needed to read one unit and write its rows.
type TableSchema struct {
	Stem          string // == main table name
	Record        protoreflect.MessageDescriptor
	Wrapper       protoreflect.MessageDescriptor
	RepeatedField protoreflect.FieldDescriptor // the repeated field inside the wrapper
	GenericField  protoreflect.FieldDescriptor // the GenericFormat field on the record
	PayloadField  protoreflect.FieldDescriptor // the payload field on the record
	TimestampFd   protoreflect.FieldDescriptor // generic_format.timestamp
	GenericFields []FieldPath
	PayloadFields []FieldPath    // scalar payload values, nested groups flattened
	Columns       []store.Column // main table columns, in row order
	OrderBy       []string
	HasRecordID   bool
	Children      []ChildSchema
	Descriptor    []byte // serialized FileDescriptorSet — portable, for workers
	SchemaVersion string
}

// Build compiles protoPath and derives the schema for a unit.
func Build(stem, protoPath string) (*TableSchema, error) {
	dir := filepath.Dir(protoPath)
	name := filepath.Base(protoPath)
	c := protocompile.Compiler{
		Resolver: protocompile.WithStandardImports(
			&protocompile.SourceResolver{ImportPaths: []string{dir}}),
		SourceInfoMode: protocompile.SourceInfoNone,
	}
	files, err := c.Compile(context.Background(), name)
	if err != nil {
		return nil, fwerr.SchemaWrap(err, "%s: failed to compile", name)
	}
	fd := files[0]

	// Serialize to a FileDescriptorSet so workers (and a future streaming
	// module) can rebuild the schema from bytes alone.
	fdset := &descriptorpb.FileDescriptorSet{}
	seen := map[string]bool{}
	var collect func(f protoreflect.FileDescriptor)
	collect = func(f protoreflect.FileDescriptor) {
		if seen[f.Path()] {
			return
		}
		seen[f.Path()] = true
		imports := f.Imports()
		for i := 0; i < imports.Len(); i++ {
			collect(imports.Get(i).FileDescriptor)
		}
		fdset.File = append(fdset.File, protodesc.ToFileDescriptorProto(f))
	}
	collect(fd)

	return buildFrom(stem, fd, fdset, name)
}

// BuildFromDescriptor rebuilds a schema from serialized descriptor bytes — the
// portable form used to hand a schema to a worker without re-reading the
// .proto from disk.
func BuildFromDescriptor(stem string, descriptorBytes []byte, source string) (*TableSchema, error) {
	fdset := &descriptorpb.FileDescriptorSet{}
	if err := protoUnmarshal(descriptorBytes, fdset); err != nil {
		return nil, fwerr.SchemaWrap(err, "%s: cannot parse descriptor set", source)
	}
	files, err := protodesc.NewFiles(fdset)
	if err != nil {
		return nil, fwerr.SchemaWrap(err, "%s: cannot build descriptors", source)
	}
	var primary protoreflect.FileDescriptor
	files.RangeFiles(func(f protoreflect.FileDescriptor) bool {
		if filepath.Base(f.Path()) == source {
			primary = f
			return false
		}
		primary = f // fallback: last file
		return true
	})
	if primary == nil {
		return nil, fwerr.Schema("%s: descriptor set is empty", source)
	}
	return buildFrom(stem, primary, fdset, source)
}

func buildFrom(stem string, fd protoreflect.FileDescriptor,
	fdset *descriptorpb.FileDescriptorSet, source string) (*TableSchema, error) {

	rec, gField, pField, wrap, repField, err := detect(fd, source)
	if err != nil {
		return nil, err
	}

	genericNames, genericCols, err := scalarColumns(gField.Message(), source)
	if err != nil {
		return nil, err
	}
	payloadNames, payloadCols, repeated, err := splitPayload(pField.Message(), source)
	if err != nil {
		return nil, err
	}

	tsFd := gField.Message().Fields().ByName("timestamp")
	if tsFd == nil {
		return nil, fwerr.Schema("%s: %s has no 'timestamp' field", source, GenericMessage)
	}

	hasChildren := len(repeated) > 0
	main := append([]store.Column{}, leadingColumns...)
	if hasChildren {
		main = append(main, recordIDColumn)
	}
	main = append(main, genericCols...)
	main = append(main, payloadCols...)
	main = append(main, trailingColumns...)

	var children []ChildSchema
	for _, f := range repeated {
		subNames, subCols, err := scalarColumns(f.Message(), source)
		if err != nil {
			return nil, err
		}
		if len(subNames) == 0 {
			return nil, fwerr.Schema("%s: payload.%s sub-message has no fields", source, f.Name())
		}
		cc := append([]store.Column{}, leadingColumns...)
		cc = append(cc, recordIDColumn)
		cc = append(cc, genericCols...)
		cc = append(cc, subCols...)
		cc = append(cc, trailingColumns...)
		if err := checkUnique(cc, fmt.Sprintf("%s:%s", source, f.Name())); err != nil {
			return nil, err
		}
		// convention: the sub-message's first field is its key dimension
		keyDim := subCols[0].Name
		children = append(children, ChildSchema{
			Table:         fmt.Sprintf("%s_%s", stem, f.Name()),
			RepeatedField: f,
			SubFields:     subNames,
			Columns:       cc,
			OrderBy:       []string{"run_id", "hostname", keyDim, "ts"},
		})
	}

	if err := checkUnique(main, source); err != nil {
		return nil, err
	}

	raw, err := protoMarshal(fdset)
	if err != nil {
		return nil, fwerr.SchemaWrap(err, "%s: cannot serialize descriptors", source)
	}
	sum := sha256.Sum256(raw)

	return &TableSchema{
		Stem:          stem,
		Record:        rec,
		Wrapper:       wrap,
		RepeatedField: repField,
		GenericField:  gField,
		PayloadField:  pField,
		TimestampFd:   tsFd,
		GenericFields: genericNames,
		PayloadFields: payloadNames,
		Columns:       main,
		OrderBy:       []string{"run_id", "hostname", "ts"},
		HasRecordID:   hasChildren,
		Children:      children,
		Descriptor:    raw,
		SchemaVersion: "sha256:" + hex.EncodeToString(sum[:])[:16],
	}, nil
}

// detect finds the record message and, if present, the wrapper message.
//
// The record message is identified by the agreed field NAMES: it has a singular
// message field named generic_format (header) and a singular message field
// named payload (data). Both are required.
func detect(fd protoreflect.FileDescriptor, source string) (
	rec protoreflect.MessageDescriptor,
	gField, pField protoreflect.FieldDescriptor,
	wrap protoreflect.MessageDescriptor,
	repField protoreflect.FieldDescriptor,
	err error) {

	msgs := fd.Messages()
	for i := 0; i < msgs.Len(); i++ {
		desc := msgs.Get(i)
		g := desc.Fields().ByName(GenericField)
		if g == nil {
			continue // not a record message
		}
		// this message carries generic_format, so it must be a valid record
		if g.Kind() != protoreflect.MessageKind || g.IsList() {
			return nil, nil, nil, nil, nil, fwerr.Schema(
				"%s: %s.%s must be a singular message field", source, desc.Name(), GenericField)
		}
		if string(g.Message().Name()) != GenericMessage {
			return nil, nil, nil, nil, nil, fwerr.Schema(
				"%s: %s.%s must be of type %s, got %s",
				source, desc.Name(), GenericField, GenericMessage, g.Message().Name())
		}
		p := desc.Fields().ByName(PayloadField)
		if p == nil {
			return nil, nil, nil, nil, nil, fwerr.Schema(
				"%s: record message %s has '%s' but no '%s' field. The contract "+
					"requires both a '%s' and a '%s' message field.",
				source, desc.Name(), GenericField, PayloadField, GenericField, PayloadField)
		}
		if p.Kind() != protoreflect.MessageKind || p.IsList() {
			return nil, nil, nil, nil, nil, fwerr.Schema(
				"%s: %s.%s must be a singular message field", source, desc.Name(), PayloadField)
		}
		if rec != nil {
			return nil, nil, nil, nil, nil, fwerr.Schema(
				"%s: more than one record message (both %s and %s have a '%s' field); expected one",
				source, rec.Name(), desc.Name(), GenericField)
		}
		rec, gField, pField = desc, g, p
	}

	if rec == nil {
		return nil, nil, nil, nil, nil, fwerr.Schema(
			"%s: no record message found. The contract requires a message with a "+
				"'%s' field (header) and a '%s' field (data).", source, GenericField, PayloadField)
	}

	// wrapper = a message with a `repeated <record>` field
	for i := 0; i < msgs.Len() && wrap == nil; i++ {
		desc := msgs.Get(i)
		for j := 0; j < desc.Fields().Len(); j++ {
			f := desc.Fields().Get(j)
			if f.IsList() && f.Kind() == protoreflect.MessageKind &&
				f.Message().FullName() == rec.FullName() {
				wrap, repField = desc, f
				break
			}
		}
	}
	return rec, gField, pField, wrap, repField, nil
}

// scalarColumns maps a message's fields to columns. Singular sub-messages are a
// 1:1 group, so they are flattened into the same row with their name as a
// prefix (io_flags.direct -> io_flags_direct). Repeated fields are rejected —
// used for the header (GenericFormat) and for a repeated sub-message's own
// fields, neither of which may fan out further.
func scalarColumns(msg protoreflect.MessageDescriptor, source string) (
	[]FieldPath, []store.Column, error) {
	return collectScalars(msg, nil, 0, source)
}

// collectScalars walks msg, descending into singular sub-messages, and returns
// one path + column per scalar found. base is the path walked so far.
//
// 1:1 data belongs on the same row, so a singular sub-message becomes prefixed
// columns rather than a child table; only a repeated message (1:N) earns a
// table of its own.
func collectScalars(msg protoreflect.MessageDescriptor, base FieldPath, depth int,
	source string) ([]FieldPath, []store.Column, error) {

	if depth > maxNestDepth {
		return nil, nil, fwerr.Schema(
			"%s: %s is nested more than %d levels deep (possible message cycle)",
			source, msg.FullName(), maxNestDepth)
	}

	var paths []FieldPath
	var cols []store.Column
	for i := 0; i < msg.Fields().Len(); i++ {
		f := msg.Fields().Get(i)
		path := append(append(FieldPath{}, base...), f)

		if f.IsMap() {
			return nil, nil, fwerr.Schema("%s: %s.%s is a map, which has no column form",
				source, msg.Name(), f.Name())
		}
		if f.IsList() {
			return nil, nil, fwerr.Schema(
				"%s: %s.%s is repeated (this message's fields must be scalars, or "+
					"singular sub-messages that flatten into the row)",
				source, msg.Name(), f.Name())
		}
		if f.Kind() == protoreflect.MessageKind || f.Kind() == protoreflect.GroupKind {
			// 1:1 group -> flatten into this row, keeping the field name as prefix
			subPaths, subCols, err := collectScalars(f.Message(), path, depth+1, source)
			if err != nil {
				return nil, nil, err
			}
			paths = append(paths, subPaths...)
			cols = append(cols, subCols...)
			continue
		}
		ct, ok := protoToType[f.Kind()]
		if !ok {
			return nil, nil, fwerr.Schema("%s: %s.%s has unsupported protobuf type %s",
				source, msg.Name(), f.Name(), f.Kind())
		}
		paths = append(paths, path)
		cols = append(cols, store.Column{Name: path.ColumnName(), Type: ct})
	}
	return paths, cols, nil
}

// splitPayload separates a payload into values that live on the record's own row
// and repeated sub-messages that become child tables.
//
//	scalar             1:1 -> a column
//	singular message   1:1 -> flattened into prefixed columns on the same row
//	repeated message   1:N -> its own child table
func splitPayload(payload protoreflect.MessageDescriptor, source string) (
	[]FieldPath, []store.Column, []protoreflect.FieldDescriptor, error) {

	var paths []FieldPath
	var cols []store.Column
	var repeated []protoreflect.FieldDescriptor

	for i := 0; i < payload.Fields().Len(); i++ {
		f := payload.Fields().Get(i)

		if f.IsList() {
			if f.Kind() != protoreflect.MessageKind {
				return nil, nil, nil, fwerr.Schema(
					"%s: payload.%s is a repeated scalar; only repeated messages "+
						"(which become child tables) are supported", source, f.Name())
			}
			repeated = append(repeated, f) // 1:N -> its own child table
			continue
		}
		if f.IsMap() {
			return nil, nil, nil, fwerr.Schema(
				"%s: payload.%s is a map, which has no column form", source, f.Name())
		}
		if f.Kind() == protoreflect.MessageKind || f.Kind() == protoreflect.GroupKind {
			// 1:1 group (e.g. io_flags) -> flatten onto this row, prefixed
			subPaths, subCols, err := collectScalars(f.Message(), FieldPath{f}, 1, source)
			if err != nil {
				return nil, nil, nil, err
			}
			paths = append(paths, subPaths...)
			cols = append(cols, subCols...)
			continue
		}
		ct, ok := protoToType[f.Kind()]
		if !ok {
			return nil, nil, nil, fwerr.Schema("%s: payload.%s has unsupported protobuf type %s",
				source, f.Name(), f.Kind())
		}
		paths = append(paths, FieldPath{f})
		cols = append(cols, store.Column{Name: string(f.Name()), Type: ct})
	}
	return paths, cols, repeated, nil
}

func checkUnique(cols []store.Column, source string) error {
	seen := map[string]bool{}
	for _, c := range cols {
		if seen[c.Name] {
			return fwerr.Schema("%s: duplicate column '%s' "+
				"(a generic and a payload/sub field share a name)", source, c.Name)
		}
		seen[c.Name] = true
	}
	return nil
}
